File: frontend/src/app/service/horse.service.ts
import {HttpClient, HttpParams} from '@angular/common/http';
import {formatDate} from '@angular/common';
import {Injectable} from '@angular/core';
import {Observable} from 'rxjs';
import {environment} from 'src/environments/environment';
import {Horse} from '../dto/horse';
import {HorseSearch} from '../dto/horse';
import {Sex} from '../dto/sex';

const baseUri = environment.backendUrl + '/horses';

@Injectable({
  providedIn: 'root'
})
export class HorseService {

  constructor(
    private http: HttpClient,
  ) { }

  /**
   * Get all horses stored in the system
   *
   * @return observable list of found horses.
   */
  getAll(): Observable<Horse[]> {
    console.log('Get all horses');
    return this.http.get<Horse[]>(baseUri + '/all');
  }

  /**
   * Get horse with the given id
   *
   * @param id of the horse
   * @return observable of received horse.
   */
  getHorseById(id: number): Observable<Horse> {
    console.log('Get horse with ' + id);
    return this.http.get<Horse>(baseUri + '/' + id);
  }

 /**
  * Search for horses by name
  *
  * @return observable list of found horses.
  */
  public searchByName(name: string, limitTo: number): Observable<Horse[]> {
     console.log('Search for horses by name');
      const params = new HttpParams()
        .set('name', name)
        .set('limit', limitTo);
      return this.http.get<Horse[]>(baseUri + '/parentsearch', { params });
    }

  /**
   * Search for horses by search
   *
   * @return observable list of found horses.
   */
   public searchHorse(horse: HorseSearch): Observable<Horse[]> {
   console.log('Search Horse' + horse);
   let params = new HttpParams();
   if(horse.name!=null){
          params = params.set('name', horse.name);
   }
   if(horse.description!=null && horse.description!==undefined && horse.description!=='') {
             params = params.set('description', horse.description);
   }
   if(horse.bornBefore!==undefined && horse.bornBefore!=null){

          const birthday = formatDate(horse.bornBefore,'yyyy-MM-dd','en-US');
          params = params.set('bornBefore', birthday);
   }
   if(horse.sex!=null && horse.sex!=='NONE'){
      params = params.set('sex', horse.sex);
   }

   if(horse.owner!=null && horse.owner!==undefined){
          if (horse.owner.id!=null && horse.owner.id!==undefined){
          params = params.set('ownerID', horse.owner.id);
          }
   }

   return this.http.get<Horse[]>(baseUri + '/search', { params });

   }

  /**
   * Create a new horse in the system.
   *
   * @param horse the data for the horse that should be created
   * @return an Observable for the created horse
   */
  create(horse: Horse): Observable<Horse> {
    console.log('Create horse: ' + horse);
    return this.http.post<Horse>(
      baseUri,
      horse
    );
  }

  /**
   * Update a new horse in the system.
   *
   * @param horse the data for the horse that should be updated
   * @return an Observable for the updated horse
   */
  update(horse: Horse, id: number): Observable<Horse> {
    console.log('Update horse: ' + horse);
    return this.http.put<Horse>(
      baseUri + '/' + id,
      horse
    );
  }

  /**
   * Delete a horse with a given id in the system.
   *
   * @param id the id of the horse that should be deleted
   */
  deleteHorse(id: number): Observable<Horse> {
     console.log('Delete horse with ID ', id);
      return this.http.delete<Horse>(baseUri + '/' + id);
  }

}

File: backend/src/main/java/at/ac/tuwien/sepm/assignment/individual/service/HorseService.java
package at.ac.tuwien.sepm.assignment.individual.service;

import at.ac.tuwien.sepm.assignment.individual.dto.HorseCreateDto;
import at.ac.tuwien.sepm.assignment.individual.dto.HorseDetailDto;
import at.ac.tuwien.sepm.assignment.individual.dto.HorseDto;
import at.ac.tuwien.sepm.assignment.individual.dto.HorseListDto;
import at.ac.tuwien.sepm.assignment.individual.dto.HorseSearchDto;
import at.ac.tuwien.sepm.assignment.individual.dto.HorseTreeDto;
import at.ac.tuwien.sepm.assignment.individual.exception.ConflictException;
import at.ac.tuwien.sepm.assignment.individual.exception.NotFoundException;
import at.ac.tuwien.sepm.assignment.individual.exception.ValidationException;

import java.util.stream.Stream;

/**
 * Service for working with horses.
 */
public interface HorseService {
  /**
   * Lists all horses stored in the system.
   *
   * @return list of all stored horses
   */
  Stream<HorseListDto> allHorses();


  /**
   * Updates the horse with the ID given in {@code horse}
   * with the data given in {@code horse}
   * in the persistent data store.
   *
   * @param horse the horse to update
   * @return he updated horse
   * @throws NotFoundException if the horse with given ID does not exist in the persistent data store
   * @throws ValidationException if the update data given for the horse is in itself incorrect (description too long, no name, …)
   * @throws ConflictException if the update data given for the horse is in conflict the data currently in the system (owner does not exist, …)
   */
  HorseDetailDto update(HorseDetailDto horse) throws NotFoundException, ValidationException, ConflictException;


  /**
   * Get the horse with given ID, with more detail information.
   * This includes the owner of the horse, and its parents.
   * The parents of the parents are not included.
   *
   * @param id the ID of the horse to get
   * @return the horse with ID {@code id}
   * @throws NotFoundException if the horse with the given ID does not exist in the persistent data store
   */
  HorseDetailDto getById(long id) throws NotFoundException;

  /**
   * Create a new horse in the persistent data store.
   *
   * @param newHorse the data for the new horse
   * @return the horse, that was just newly created in the persistent data store
   * @throws ValidationException if the parameters are not allowed
   */
  HorseDto create(HorseCreateDto newHorse) throws ValidationException, ConflictException;

  /**
   * Search for horses matching the criteria in {@code searchParameters}.
   * <p>
   * A horse is considered matched, if its name contains {@code searchParameters.name} as a substring.
   * The returned stream of horses never contains more than {@code searchParameters.maxAmount} elements,
   *  even if there would be more matches in the persistent data store.
   * </p>
   *
   * @param searchParameters object containing the search parameters to match
   * @return a stream containing owners matching the criteria in {@code searchParameters}
   */
  Stream<HorseDto> searchByName(HorseSearchDto searchParameters);

  /**
   * Delete the horse with given ID.
   * Parents/child relationships gets removed once a horse is deleted.
   *
   * @param id the ID of the horse to delete
   * @throws NotFoundException if the horse with the given ID does not exist in the persistent data store
   */
  void deleteHorse(long id) throws NotFoundException;

  /**
   * Search for horses matching the criteria in {@code searchParameters}.
   * <p>
   * A horse is considered matched, if its values contain the search parameters as a substring. The search is
   * case insensitive. The possible values are {@code searchParameters.name}, {@code searchParameters.description},
   * {@code searchParameters.bornBefore}, {@code searchParameters.sex} and {@code searchParameters.ownerName}.
   * </p>
   *
   * @param searchParameters object containing the search parameters to match
   * @return a stream containing owners matching the criteria in {@code searchParameters}
   */
  Stream<HorseListDto> searchHorses(HorseSearchDto searchParameters);

  /**
   * Get the horse with given ID.
   *
   * @param id the ID of the horse to get
   * @param generations number of horse generations to get the horses from
   * @return the horses with ID {@code id} and their parent horses within the choosen generations
   * @throws NotFoundException if the horse with the given ID does not exist in the persistent data store
   */
  HorseTreeDto getHorseTree(Long id, Long generations) throws NotFoundException;
}
